module/zfs/zio.c

```diff
@@ -3326,8 +3346,48 @@ zio_ddt_free(zio_t *zio)
 	freedde = dde = ddt_lookup(ddt, bp, B_TRUE);
 	if (dde) {
 		ddp = ddt_phys_select(dde, bp);
-		if (ddp)
-			ddt_phys_decref(ddp);
+	}
+	if (ddp != NULL) {
+		ddt_phys_decref(ddp);
+	} else {
```

Comment:
I'm not completely following the cleanup here tbh. Would it be easier to not change `ddt_lookup()` and instead mark the dde as one that we do not want to sync? Or maybe you could explain what's going on here some more in the comments.

Reply:
This came from Matt's code (related to the ddt-log project), as I recall.

Reply:
looks like the below comment needs to be updated as Matt referenced